src/Features/Core/Portable/EmbeddedLanguages/DateTime/DateTimeEmbeddedCompletionProvider.cs

```diff
@@ -0,0 +1,219 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.DateTime;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Features.EmbeddedLanguages.DateTime
+{
+    internal partial class DateTimeEmbeddedCompletionProvider : CompletionProvider
+    {
+        private const string StartKey = nameof(StartKey);
+        private const string LengthKey = nameof(LengthKey);
+        private const string NewTextKey = nameof(NewTextKey);
+        private const string NewPositionKey = nameof(NewPositionKey);
+        private const string DescriptionKey = nameof(DescriptionKey);
+
+        // Always soft-select these completion items.  Also, never filter down.
+        private static readonly CompletionItemRules s_rules =
+            CompletionItemRules.Default.WithSelectionBehavior(CompletionItemSelectionBehavior.SoftSelection)
+                                       .WithFilterCharacterRule(CharacterSetModificationRule.Create(CharacterSetModificationKind.Replace, new char[] { }));
+
+        private readonly DateTimeEmbeddedLanguageFeatures _language;
+
+        public DateTimeEmbeddedCompletionProvider(DateTimeEmbeddedLanguageFeatures language)
+        {
+            _language = language;
+        }
+
+        public override bool ShouldTriggerCompletion(SourceText text, int caretPosition, CompletionTrigger trigger, OptionSet options)
+        {
+            if (trigger.Kind == CompletionTriggerKind.Invoke ||
+                trigger.Kind == CompletionTriggerKind.InvokeAndCommitIfUnique)
+            {
+                return true;
+            }
+
+            if (trigger.Kind == CompletionTriggerKind.Insertion)
+            {
+                return char.IsLetter(trigger.Character);
+            }
+
+            return false;
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            if (!context.Options.GetOption(DateTimeOptions.ProvideDateTimeOptionsCompletions, context.Document.Project.Language))
```

Comment:
Is an option necessary?  As far as I can tell there were no completions before, so we wouldn't be changing muscle memory right since it's soft selected?

If people get really annoyed an option could be added later on

Reply:
this is also an easy escape hatch for us.  But i agree.  we likely don't need a user visible UI option for this.